### PR TITLE
[SYCL-MLIR][NFC][LICM] Add `LICMCandidate` class 

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Add `LICMCandidate` class to indicate all the other operations an operation depends on to be hoisted.
It will also be used to store version conditions needed for the operation to be hoisted.

### CODE REVIEW

**Review by victor-eds — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 1]
```diff
@@ -703,16 +703,44 @@ void AffineParallelGuardBuilder::getConstraints(
...
 
+//===----------------------------------------------------------------------===//
+// LICMCandidate
+//===----------------------------------------------------------------------===//
+
+class LICMCandidate {
+public:
+  LICMCandidate(Operation &op) : op(op) {}
+  Operation &getOp() const { return op; }
+  void addDependee(Operation &op) { dependees.push_back(&op); }
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:714]: How about we call this member function `addPrerequisite` ?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 2]
```diff
@@ -703,16 +703,44 @@ void AffineParallelGuardBuilder::getConstraints(
...
+
+private:
+  Operation &op;
+  SmallVector<Operation *> dependees;
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:718]: `dependees` -> prerequisites; // Operations that need to be hoisted to allow this operation to be hoisted.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 3]
```diff
@@ -703,16 +703,44 @@ void AffineParallelGuardBuilder::getConstraints(
...
+};
+
 //===----------------------------------------------------------------------===//
 
+/// Check if \p willBeMoved contains \p op. If true, add \p op to \p candidate
+/// list of dependees, and return true. Otherwise, return false.
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:724]: If `op` is an operation that has already been proven invariant (and therefore is in `willBeMoved`) what makes it a `dependee` of `candidate` ? To understand one requires the content and given this function is quite small I prefer to `inline` it.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 4]
```diff
@@ -703,16 +703,44 @@ void AffineParallelGuardBuilder::getConstraints(
...
 }
 
+//===----------------------------------------------------------------------===//
+// LICMCandidate
+//===----------------------------------------------------------------------===//
+
+class LICMCandidate {
+public:
+  LICMCandidate(Operation &op) : op(op) {}
+  Operation &getOp() const { return op; }
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:713]: Change name `getOp` to `getOperation` (I see that `OperationSideEffects` in the same file uses that term` already).

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 5]
```diff
@@ -703,16 +703,44 @@ void AffineParallelGuardBuilder::getConstraints(
...
+  void addDependee(Operation &op) { dependees.push_back(&op); }
+
+private:
+  Operation &op;
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:717]: next PR.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 6]
```diff
@@ -703,16 +703,44 @@ void AffineParallelGuardBuilder::getConstraints(
...
+//===----------------------------------------------------------------------===//
+// LICMCandidate
+//===----------------------------------------------------------------------===//
+
+class LICMCandidate {
+public:
+  LICMCandidate(Operation &op) : op(op) {}
+  Operation &getOp() const { return op; }
+  void addDependee(Operation &op) { dependees.push_back(&op); }
+
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:715]: Will add when there is a user.

**Pull request merged**